### DISCUSSION TITLE
ORAS v1.3.0 milestone discussion

### DISCUSSION THREAD

**Discussion by FeynmanZhou** [oras-project/oras#1311] [General]
Hi ORAS fellows:

ORAS continues to follow a quarter-based iterative release cadence in 2024. ORAS v1.2.0-beta.1 was released on Feb 7 and is going to be stable in May (updated: released on May 30). 

It's time to discuss and determine the v1.3.0 milestone and release plan for v1.2.0 in the community. After regularly triaged all GitHub issues, I grouped the major enhancements into four aspects as follows. 

This is an open discussion and I am also open to any suggestions. Feel free to share anything feature requests you think are missing below or should be prioritized.

## Improve verbose logs for all ORAS commands to enhance the troubleshooting experience

Make the verbose logs consistent, useful, concise, and human-readable. Users can easily locate the root cause by analyzing the verbose logs. Formulize the CLI output guideline to provide a unified reference for future command output.

   - https://github.com/oras-project/oras/issues/911
   - https://github.com/oras-project/oras/issues/1091
   - https://github.com/oras-project/oras/issues/916

Part of its enhancements could be considered in ORAS v1.2.1 if possible.

## Support multi-arch image index management 

Most of the open-source projects support multi-arch images. Enable users to push/attach a multi-arch image or an image index by leveraging the [OCI image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md).

- https://github.com/oras-project/oras/issues/1066
- https://github.com/oras-project/oras/issues/1053
- https://github.com/oras-project/oras/issues/813

This will be marked as a preview feature when we introduce a brand-new feature

## Extend functionality of oras pull/push

These are frequently asked feature requests so it would be good to define the specifications in v1.3.0 at least. 

- https://github.com/oras-project/oras/issues/1160 and another duplicated https://github.com/oras-project/oras/issues/1366
- https://github.com/oras-project/oras/issues/1308

## Streamline annotating experience (nice to have)

Annotating experience is critical to allow users to update the OCI image annotation metadata. For example, in the containers secure supply chain, users can easily update the image lifecycle metadata without learning the OCI Spec. such as adding "End-of-support" metadata to an OCI image manifest. 

In addition, ORAS requires a static JSON file as an input to update annotations in some cases. This is not convenient in automation and scripting scenario.

- https://github.com/oras-project/oras/issues/571

This will be marked as a preview feature when we introduce a brand-new feature

## Improve user experience and performance

- https://github.com/oras-project/oras/issues/491
- https://github.com/oras-project/oras/issues/1108
- https://github.com/oras-project/oras/issues/742
- https://github.com/oras-project/oras/issues/1061

**Comment by FeynmanZhou** [comment]
As we discussed in the [community meeting on Mar 26, 2024](https://hackmd.io/P-O6n222TcSMoJgHmTTduw?both#Notes), @sajayantony suggested consider another two enhancements for `oras pull`. It would be better that we could define specifications in ORAS v1.3.0 for these two:

- https://github.com/oras-project/oras/issues/1160
- https://github.com/oras-project/oras/issues/1308

**Comment by TerryHowe** [comment]
Looks good, I just added a comment to one of the issues.
